pkg/controller/build/build_controller.go

```diff
@@ -1174,206 +1218,277 @@ func (ctrl *Controller) handleCanonicalizedPullSecret(secret *corev1.Secret) (*c
 	}
 	return err
 }
 
-// Determine if we have a config change.
-func isPoolConfigChange(oldPool, curPool *mcfgv1.MachineConfigPool) bool {
-	return oldPool.Spec.Configuration.Name != curPool.Spec.Configuration.Name
+func configChangeCauseBuild(old, cur *mcfgv1alpha1.MachineOSConfig) bool {
+	return equality.Semantic.DeepEqual(old.Spec.BuildInputs, cur.Spec.BuildInputs)
 }
 
-// Checks our pool to see if we can do a build. We base this off of a few criteria:
-// 1. Is the pool opted into layering?
-// 2. Do we have an object reference to an in-progress build?
-// 3. Is the pool degraded?
-// 4. Is our build in a specific state?
-//
-// Returns true if we are able to build.
-func canPoolBuild(ps *poolState) bool {
-	// If we don't have a layered pool, we should not build.
-	if !ps.IsLayered() {
-		return false
-	}
+// Determines if we should do a build based upon the state of our
+// MachineConfigPool, the presence of a build pod, etc.
+func shouldWeDoABuild(builder interface {
+	IsBuildRunning(*mcfgv1alpha1.MachineOSBuild, *mcfgv1alpha1.MachineOSConfig) (bool, error)
+}, mosc *mcfgv1alpha1.MachineOSConfig, oldMOSB, curMOSB *mcfgv1alpha1.MachineOSBuild) (bool, error) {
+	// get desired and current. If desired != current,
+	// assume we are doing a build. remove the whole layered pool annotation workflow
 
-	// If we have a reference to an in-progress build, we should not build.
-	if ps.HasBuildObjectForCurrentMachineConfig() {
-		return false
-	}
+	if oldMOSB.Spec.DesiredConfig != curMOSB.Spec.DesiredConfig {
+		// the desiredConfig changed. We need to do an update
+		// but check that there isn't already a build.
+		// If a build is found running, we should not do a build.
+		isRunning, err := builder.IsBuildRunning(curMOSB, mosc)
 
-	// If the pool is degraded, we should not build.
-	if ps.IsAnyDegraded() {
-		return false
-	}
+		return !isRunning, err
 
-	// If the pool is in any of these states, we should not build.
-	if ps.IsBuilding() {
-		return false
+		// check for image pull sped changing?
 	}
+	return false, nil
+}
 
-	if ps.IsBuildPending() {
-		return false
+// Determines if a pod or build is managed by this controller by examining its labels.
+func hasAllRequiredOSBuildLabels(labels map[string]string) bool {
+	requiredLabels := []string{
+		ctrlcommon.OSImageBuildPodLabel,
+		targetMachineConfigPoolLabel,
+		desiredConfigLabel,
 	}
 
-	if ps.IsBuildFailure() {
-		return false
+	for _, label := range requiredLabels {
+		if _, ok := labels[label]; !ok {
+			return false
+		}
 	}
 
 	return true
 }
 
-// Determines if we should do a build based upon the state of our
-// MachineConfigPool, the presence of a build pod, etc.
-func shouldWeDoABuild(builder interface {
-	IsBuildRunning(*mcfgv1.MachineConfigPool) (bool, error)
-}, oldPool, curPool *mcfgv1.MachineConfigPool) (bool, error) {
-	ps := newPoolState(curPool)
+func (ctrl *Controller) doesMOSBExist(mosc *mcfgv1alpha1.MachineOSConfig) (*mcfgv1alpha1.MachineOSBuild, bool) {
```

Comment:
praise: Nicely done with this!

We'll probably want to make a more canonical shared implementation of this in the future since anything that works with MachineOSConfigs and MachineOSBuilds will need to perform similar operations.